File: src/net/dinkla/raytracer/objects/beveled/BeveledBox.java
package net.dinkla.raytracer.objects.beveled;

import net.dinkla.raytracer.math.BBox;
import net.dinkla.raytracer.math.Normal;
import net.dinkla.raytracer.math.Point3D;
import net.dinkla.raytracer.math.Vector3D;
import net.dinkla.raytracer.objects.*;
import net.dinkla.raytracer.objects.compound.Compound;

/**
 * Created by IntelliJ IDEA.
 * User: jorndinkla
 * Date: 26.05.2010
 * Time: 22:46:43
 * To change this template use File | Settings | File Templates.
 */
public class BeveledBox extends Compound {

    public final Point3D p0;
    public final Point3D p1;
    public final float rb;
    public BBox bbox;

    public BeveledBox(final Point3D p0, final Point3D p1, final float rb, boolean isWiredFrame) {
        super();

        this.p0 = p0;
        this.p1 = p1;
        this.rb = rb;

        Instance top_front_edge = new Instance (new OpenCylinder(-(p1.getX() - p0.getX() - 2 * rb) / 2, (p1.getX() - p0.getX() - 2 * rb) / 2, rb));	// top front edge
        top_front_edge.rotateZ(90);
        top_front_edge.translate((p0.getX() + p1.getX()) / 2, p1.getY() - rb, p1.getZ() - rb);
        //top_front_edge.transform_texture(false);
        objects.add(top_front_edge);

        // top back (-ve z)

        Instance top_back_edge = new Instance (new OpenCylinder(-(p1.getX() - p0.getX() - 2 * rb) / 2, (p1.getX() - p0.getX() - 2 * rb) / 2, rb));	// top back edge
        top_back_edge.rotateZ(90);
        top_back_edge.translate((p0.getX() + p1.getX()) / 2, p1.getY() - rb, p0.getZ() + rb);
        //top_back_edge->transform_texture(false);
        objects.add(top_back_edge);


        // top right (+ve x)

        Instance top_right_edge = new Instance (new OpenCylinder(-(p1.getZ() - p0.getZ() - 2 * rb) / 2, (p1.getZ() - p0.getZ() - 2 * rb) / 2, rb)); // top right edge
        top_right_edge.rotateX(90);
        top_right_edge.translate(p1.getX() - rb, p1.getY() - rb, (p0.getZ() + p1.getZ()) / 2);
        //top_right_edge->transform_texture(false);
        objects.add(top_right_edge);


        // top left (-ve x)

        Instance top_left_edge = new Instance (new OpenCylinder(-(p1.getZ() - p0.getZ() - 2 * rb) / 2, (p1.getZ() - p0.getZ() - 2 * rb) / 2, rb)); // top left edge
        top_left_edge.rotateX(90);
        top_left_edge.translate(p0.getX() + rb, p1.getY() - rb, (p0.getZ() + p1.getZ()) / 2);
        //top_left_edge->transform_texture(false);
        objects.add(top_left_edge);

        // bottom edges  (-ve y)

        // bottom front  (+ve z)

        Instance bottom_front_edge = new Instance (new OpenCylinder(-(p1.getX() - p0.getX() - 2 * rb) / 2, (p1.getX() - p0.getX() - 2 * rb) / 2, rb));	// bottom fromt edge
        bottom_front_edge.rotateZ(90);
        bottom_front_edge.translate((p0.getX() + p1.getX()) / 2, p0.getY() + rb, p1.getZ() - rb);
        //bottom_front_edge->transform_texture(false);
        objects.add(bottom_front_edge);


        // bottom back  (-ve z)

        Instance bottom_back_edge = new Instance (new OpenCylinder(-(p1.getX() - p0.getX() - 2 * rb) / 2, (p1.getX() - p0.getX() - 2 * rb) / 2, rb));	// bottom back edge
        bottom_back_edge.rotateZ(90);
        bottom_back_edge.translate((p0.getX() + p1.getX()) / 2, p0.getY() + rb, p0.getZ() + rb);
        //bottom_back_edge->transform_texture(false);
        objects.add(bottom_back_edge);


        // bottom right (-ve x, -ve y)

        Instance bottom_right_edge = new Instance (new OpenCylinder(-(p1.getZ() - p0.getZ() - 2 * rb) / 2, (p1.getZ() - p0.getZ() - 2 * rb) / 2, rb)); // bottom right edge
        bottom_right_edge.rotateX(90);
        bottom_right_edge.translate(p1.getX() - rb, p0.getY() + rb, (p0.getZ() + p1.getZ()) / 2);
        //bottom_right_edge->transform_texture(false);
        objects.add(bottom_right_edge);

        // bottom left (-ve x, -ve y)

        Instance bottom_left_edge = new Instance (new OpenCylinder(-(p1.getZ() - p0.getZ() - 2 * rb) / 2, (p1.getZ() - p0.getZ() - 2 * rb) / 2, rb)); // bottom left edge
        bottom_left_edge.rotateX(90);
        bottom_left_edge.translate(p0.getX() + rb, p0.getY() + rb, (p0.getZ() + p1.getZ()) / 2);
        //bottom_left_edge->transform_texture(false);
        objects.add(bottom_left_edge);


        // vertical edges

        // vertical right front  (+ve x, +ve z)

        Instance vertical_right_front_edge = new Instance (new OpenCylinder(p0.getY() + rb, p1.getY() - rb, rb));
        vertical_right_front_edge.translate(p1.getX() - rb, 0, p1.getZ() - rb);
        //vertical_right_front_edge->transform_texture(false);
        objects.add(vertical_right_front_edge);

        // vertical left front  (-ve x, +ve z)

        Instance vertical_left_front_edge = new Instance (new OpenCylinder(p0.getY() + rb, p1.getY() - rb, rb));
        vertical_left_front_edge.translate(p0.getX() + rb, 0, p1.getZ() - rb);
        //vertical_left_front_edge->transform_texture(false);
        objects.add(vertical_left_front_edge);

        // vertical left back  (-ve x, -ve z)

        Instance vertical_left_back_edge = new Instance (new OpenCylinder(p0.getY() + rb, p1.getY() - rb, rb));
        vertical_left_back_edge.translate(p0.getX() + rb, 0, p0.getZ() + rb);
        //vertical_left_back_edge->transform_texture(false);
        objects.add(vertical_left_back_edge);


        // vertical right back  (+ve x, -ve z)

        Instance vertical_right_back_edge = new Instance (new OpenCylinder(p0.getY() + rb, p1.getY() - rb, rb));
        vertical_right_back_edge.translate(p1.getX() - rb, 0, p0.getZ() + rb);
        //vertical_right_back_edge->transform_texture(false);
        objects.add(vertical_right_back_edge);

        // corner spheres

        // top right front

        Sphere top_right_front_corner = new Sphere(new Point3D(p1.getX() - rb, p1.getY() - rb, p1.getZ() - rb), rb);
        objects.add(top_right_front_corner);

        // top left front  (-ve x)

        Sphere top_left_front_corner = new Sphere(new Point3D(p0.getX() + rb, p1.getY() - rb, p1.getZ() - rb), rb);
        objects.add(top_left_front_corner);

        // top left back

        Sphere top_left_back_corner = new Sphere(new Point3D(p0.getX() + rb, p1.getY() - rb, p0.getZ() + rb), rb);
        objects.add(top_left_back_corner);

        // top right back

        Sphere top_right_back_corner = new Sphere(new Point3D(p1.getX() - rb, p1.getY() - rb, p0.getZ() + rb), rb);
        objects.add(top_right_back_corner);

        // bottom right front

        Sphere bottom_right_front_corner = new Sphere(new Point3D(p1.getX() - rb, p0.getY() + rb, p1.getZ() - rb), rb);
        objects.add(bottom_right_front_corner);

        // bottom left front

        Sphere bottom_left_front_corner = new Sphere(new Point3D(p0.getX() + rb, p0.getY() + rb, p1.getZ() - rb), rb);
        objects.add(bottom_left_front_corner);

        // bottom left back

        Sphere bottom_left_back_corner = new Sphere(new Point3D(p0.getX() + rb, p0.getY() + rb, p0.getZ() + rb), rb);
        objects.add(bottom_left_back_corner);

        // bottom right back

        Sphere bottom_right_back_corner = new Sphere(new Point3D(p1.getX() - rb, p0.getY() + rb, p0.getZ() + rb), rb);
        objects.add(bottom_right_back_corner);


        // the faces

        // bottom face: -ve y

        if (isWiredFrame) {
            return;
        }
        
        Rectangle bottom_face = new Rectangle(new Point3D(p0.getX() + rb, p0.getY(), p0.getZ() + rb),
                                              new Vector3D(0, 0, (p1.getZ() - rb) - (p0.getZ() + rb)),
                                              new Vector3D((p1.getX() - rb) - (p0.getX() + rb), 0, 0),
                                              new Normal(0, -1, 0));
        objects.add(bottom_face);


        // bottom face: +ve y

        Rectangle top_face = new Rectangle(new Point3D(p0.getX() + rb, p1.getY(), p0.getZ() + rb),
                                           new Vector3D(0, 0, (p1.getZ() - rb) - (p0.getZ() + rb)),
                                           new Vector3D((p1.getX() - rb) - (p0.getX() + rb), 0, 0),
                                           new Normal(0, 1, 0));
        objects.add(top_face);


        // back face: -ve z

        Rectangle back_face 	= new Rectangle(new Point3D(p0.getX() + rb, p0.getY() + rb, p0.getZ()),
                                                new Vector3D((p1.getX() - rb) - (p0.getX() + rb), 0, 0),
                                                new Vector3D(0, (p1.getY() - rb) - (p0.getY() + rb), 0),
                                                new Normal(0, 0 , -1));
        objects.add(back_face);


        // front face: +ve z

        Rectangle front_face 	= new Rectangle(new Point3D(p0.getX() + rb, p0.getY() + rb, p1.getZ()),
                                                new Vector3D((p1.getX() - rb) - (p0.getX() + rb), 0, 0),
                                                new Vector3D(0, (p1.getY() - rb) - (p0.getY() + rb), 0),
                                                new Normal(0, 0 , 1));
        objects.add(front_face);


        // left face: -ve x

        Rectangle left_face 	= new Rectangle(new Point3D(p0.getX(), p0.getY() + rb, p0.getZ() + rb),
                                                new Vector3D(0, 0, (p1.getZ() - rb) - (p0.getZ() + rb)),
                                                new Vector3D(0, (p1.getY() - rb) - (p0.getY() + rb), 0),
                                                new Normal(-1, 0 , 0));
        objects.add(left_face);


        // right face: +ve x

        Rectangle right_face 	= new Rectangle(new Point3D(p1.getX(), p0.getY() + rb, p0.getZ() + rb),
                                                new Vector3D(0, 0, (p1.getZ() - rb) - (p0.getZ() + rb)),
                                                new Vector3D(0, (p1.getY() - rb) - (p0.getY() + rb), 0),
                                                new Normal(1, 0 , 0));
        objects.add(right_face);
    }
    
}

File: src/main/kotlin/net/dinkla/raytracer/objects/beveled/BeveledBox.kt
package net.dinkla.raytracer.objects.beveled

import net.dinkla.raytracer.math.Normal
import net.dinkla.raytracer.math.Point3D
import net.dinkla.raytracer.math.Vector3D
import net.dinkla.raytracer.objects.Instance
import net.dinkla.raytracer.objects.OpenCylinder
import net.dinkla.raytracer.objects.Rectangle
import net.dinkla.raytracer.objects.Sphere
import net.dinkla.raytracer.objects.compound.Compound
import net.dinkla.raytracer.utilities.equals
import net.dinkla.raytracer.interfaces.hash

class BeveledBox(val p0: Point3D,
                 val p1: Point3D,
                 private val rb: Double,
                 private val isWiredFrame: Boolean = false) : Compound() {
    init {

        val topFrontEdge = Instance(OpenCylinder(-(p1.x - p0.x - 2 * rb) / 2, (p1.x - p0.x - 2 * rb) / 2, rb))    // top front edge
        topFrontEdge.rotateZ(90.0)
        topFrontEdge.translate((p0.x + p1.x) / 2, p1.y - rb, p1.z - rb)
        //top_front_edge.transform_texture(false);
        objects.add(topFrontEdge)

        // top back (-ve z)
        val topBackEdge = Instance(OpenCylinder(-(p1.x - p0.x - 2 * rb) / 2, (p1.x - p0.x - 2 * rb) / 2, rb))    // top back edge
        topBackEdge.rotateZ(90.0)
        topBackEdge.translate((p0.x + p1.x) / 2, p1.y - rb, p0.z + rb)
        //top_back_edge->transform_texture(false);
        objects.add(topBackEdge)


        // top right (+ve x)
        val topRightEdge = Instance(OpenCylinder(-(p1.z - p0.z - 2 * rb) / 2, (p1.z - p0.z - 2 * rb) / 2, rb)) // top right edge
        topRightEdge.rotateX(90.0)
        topRightEdge.translate(p1.x - rb, p1.y - rb, (p0.z + p1.z) / 2)
        //top_right_edge->transform_texture(false);
        objects.add(topRightEdge)


        // top left (-ve x)
        val topLeftEdge = Instance(OpenCylinder(-(p1.z - p0.z - 2 * rb) / 2, (p1.z - p0.z - 2 * rb) / 2, rb)) // top left edge
        topLeftEdge.rotateX(90.0)
        topLeftEdge.translate(p0.x + rb, p1.y - rb, (p0.z + p1.z) / 2)
        //top_left_edge->transform_texture(false);
        objects.add(topLeftEdge)

        // bottom edges  (-ve y)

        // bottom front  (+ve z)
        val bottomFrontEdge = Instance(OpenCylinder(-(p1.x - p0.x - 2 * rb) / 2, (p1.x - p0.x - 2 * rb) / 2, rb))    // bottom fromt edge
        bottomFrontEdge.rotateZ(90.0)
        bottomFrontEdge.translate((p0.x + p1.x) / 2, p0.y + rb, p1.z - rb)
        //bottom_front_edge->transform_texture(false);
        objects.add(bottomFrontEdge)


        // bottom back  (-ve z)
        val bottomBackEdge = Instance(OpenCylinder(-(p1.x - p0.x - 2 * rb) / 2, (p1.x - p0.x - 2 * rb) / 2, rb))    // bottom back edge
        bottomBackEdge.rotateZ(90.0)
        bottomBackEdge.translate((p0.x + p1.x) / 2, p0.y + rb, p0.z + rb)
        //bottom_back_edge->transform_texture(false);
        objects.add(bottomBackEdge)


        // bottom right (-ve x, -ve y)

        val bottomRightEdge = Instance(OpenCylinder(-(p1.z - p0.z - 2 * rb) / 2, (p1.z - p0.z - 2 * rb) / 2, rb)) // bottom right edge
        bottomRightEdge.rotateX(90.0)
        bottomRightEdge.translate(p1.x - rb, p0.y + rb, (p0.z + p1.z) / 2)
        //bottom_right_edge->transform_texture(false);
        objects.add(bottomRightEdge)

        // bottom left (-ve x, -ve y)

        val bottomLeftEdge = Instance(OpenCylinder(-(p1.z - p0.z - 2 * rb) / 2, (p1.z - p0.z - 2 * rb) / 2, rb)) // bottom left edge
        bottomLeftEdge.rotateX(90.0)
        bottomLeftEdge.translate(p0.x + rb, p0.y + rb, (p0.z + p1.z) / 2)
        //bottom_left_edge->transform_texture(false);
        objects.add(bottomLeftEdge)


        // vertical edges

        // vertical right front  (+ve x, +ve z)

        val verticalRightFrontEdge = Instance(OpenCylinder(p0.y + rb, p1.y - rb, rb))
        verticalRightFrontEdge.translate(p1.x - rb, 0.0, p1.z - rb)
        //vertical_right_front_edge->transform_texture(false);
        objects.add(verticalRightFrontEdge)

        // vertical left front  (-ve x, +ve z)

        val verticalLeftFrontEdge = Instance(OpenCylinder(p0.y + rb, p1.y - rb, rb))
        verticalLeftFrontEdge.translate(p0.x + rb, 0.0, p1.z - rb)
        //vertical_left_front_edge->transform_texture(false);
        objects.add(verticalLeftFrontEdge)

        // vertical left back  (-ve x, -ve z)

        val verticalLeftBackEdge = Instance(OpenCylinder(p0.y + rb, p1.y - rb, rb))
        verticalLeftBackEdge.translate(p0.x + rb, 0.0, p0.z + rb)
        //vertical_left_back_edge->transform_texture(false);
        objects.add(verticalLeftBackEdge)


        // vertical right back  (+ve x, -ve z)

        val verticalRightBackEdge = Instance(OpenCylinder(p0.y + rb, p1.y - rb, rb))
        verticalRightBackEdge.translate(p1.x - rb, 0.0, p0.z + rb)
        //vertical_right_back_edge->transform_texture(false);
        objects.add(verticalRightBackEdge)

        // corner spheres

        // top right front

        val topRightFrontCorner = Sphere(Point3D(p1.x - rb, p1.y - rb, p1.z - rb), rb)
        objects.add(topRightFrontCorner)

        // top left front  (-ve x)

        val topLeftFrontCorner = Sphere(Point3D(p0.x + rb, p1.y - rb, p1.z - rb), rb)
        objects.add(topLeftFrontCorner)

        // top left back

        val topLeftBackCorner = Sphere(Point3D(p0.x + rb, p1.y - rb, p0.z + rb), rb)
        objects.add(topLeftBackCorner)

        // top right back

        val topRightBackCorner = Sphere(Point3D(p1.x - rb, p1.y - rb, p0.z + rb), rb)
        objects.add(topRightBackCorner)

        // bottom right front

        val bottomRightFrontCorner = Sphere(Point3D(p1.x - rb, p0.y + rb, p1.z - rb), rb)
        objects.add(bottomRightFrontCorner)

        // bottom left front

        val bottomLeftFrontCorner = Sphere(Point3D(p0.x + rb, p0.y + rb, p1.z - rb), rb)
        objects.add(bottomLeftFrontCorner)

        // bottom left back

        val bottomLeftBackCorner = Sphere(Point3D(p0.x + rb, p0.y + rb, p0.z + rb), rb)
        objects.add(bottomLeftBackCorner)

        // bottom right back

        val bottomRightBackCorner = Sphere(Point3D(p1.x - rb, p0.y + rb, p0.z + rb), rb)
        objects.add(bottomRightBackCorner)

        // the faces
        if (!isWiredFrame) {
            // bottom face: -ve y
            val bottomFace = Rectangle(Point3D(p0.x + rb, p0.y, p0.z + rb),
                    Vector3D(0.0, 0.0, p1.z - rb - (p0.z + rb)),
                    Vector3D(p1.x - rb - (p0.x + rb), 0.0, 0.0),
                    Normal.DOWN)
            objects.add(bottomFace)

            // bottom face: +ve y
            val topFace = Rectangle(Point3D(p0.x + rb, p1.y, p0.z + rb),
                    Vector3D(0.0, 0.0, p1.z - rb - (p0.z + rb)),
                    Vector3D(p1.x - rb - (p0.x + rb), 0.0, 0.0),
                    Normal.UP)
            objects.add(topFace)

            // back face: -ve z
            val backFace = Rectangle(Point3D(p0.x + rb, p0.y + rb, p0.z),
                    Vector3D(p1.x - rb - (p0.x + rb), 0.0, 0.0),
                    Vector3D(0.0, p1.y - rb - (p0.y + rb), 0.0),
                    Normal.BACKWARD)
            objects.add(backFace)

            // front face: +ve z
            val frontFace = Rectangle(Point3D(p0.x + rb, p0.y + rb, p1.z),
                    Vector3D(p1.x - rb - (p0.x + rb), 0.0, 0.0),
                    Vector3D(0.0, p1.y - rb - (p0.y + rb), 0.0),
                    Normal.FORWARD)
            objects.add(frontFace)

            // left face: -ve x
            val leftFace = Rectangle(Point3D(p0.x, p0.y + rb, p0.z + rb),
                    Vector3D(0.0, 0.0, p1.z - rb - (p0.z + rb)),
                    Vector3D(0.0, p1.y - rb - (p0.y + rb), 0.0),
                    Normal.LEFT)
            objects.add(leftFace)

            // right face: +ve x
            val rightFace = Rectangle(Point3D(p1.x, p0.y + rb, p0.z + rb),
                    Vector3D(0.0, 0.0, p1.z - rb - (p0.z + rb)),
                    Vector3D(0.0, p1.y - rb - (p0.y + rb), 0.0),
                    Normal.RIGHT)
            objects.add(rightFace)
        }
    }

    override fun equals(other: Any?): Boolean = this.equals<BeveledBox>(other) { a, b ->
        a.p0 == b.p0 && a.p1 == b.p1 && a.rb == b.rb && a.isWiredFrame == b.isWiredFrame
    }

    override fun hashCode(): Int = hash(p0, p1, rb, isWiredFrame)

    override fun toString(): String = "BeveledBox($p0, $p1, $rb, $isWiredFrame)"
}
